File: frontend/ui/src/app/types/deployment-timeline.ts
import {DeploymentPlan} from './deployment-plan';
import {ReleaseBundleComponentType} from './release-bundle';

export type DeploymentTaskStatus = 'QUEUED' | 'RUNNING' | 'SUCCEEDED' | 'FAILED' | 'CANCELED';
export type DeploymentStepRunStatus = 'PENDING' | 'RUNNING' | 'SUCCEEDED' | 'FAILED' | 'SKIPPED';
export type DeploymentStepRunEventType = 'STARTED' | 'PROGRESS' | 'LOG' | 'OUTPUT' | 'SUCCEEDED' | 'FAILED';
export type DeploymentTimelineChangeKind = 'unchanged' | 'added' | 'removed' | 'changed';
export type DeploymentTimelineItemSource = 'task' | 'legacy_deployment';

export interface DeploymentTimelineQuery {
  applicationId?: string;
  releaseBundleId?: string;
  environmentId?: string;
  deploymentTargetId?: string;
  customerOrganizationId?: string;
  limit?: number;
  includeNonTerminal?: boolean;
}

export interface DeploymentTimeline {
  items: DeploymentTimelineItem[];
}

export interface DeploymentCompatibilityAvailability {
  processSnapshot: boolean;
  variableSnapshot: boolean;
  channel: boolean;
  environment: boolean;
  taskLogs: boolean;
  redeployPlan: boolean;
}

export interface DeploymentTimelineItem {
  source?: DeploymentTimelineItemSource;
  taskId?: string;
  legacyDeploymentId?: string;
  legacyDeploymentRevisionId?: string;
  syntheticReleaseId?: string;
  deploymentPlanId?: string;
  deploymentPlanTargetId?: string;
  deploymentTargetId: string;
  applicationId: string;
  applicationName: string;
  releaseBundleId?: string;
  releaseNumber: string;
  channelId?: string;
  channelName: string;
  environmentId?: string;
  environmentName: string;
  customerOrganizationId?: string;
  deploymentTargetName: string;
  actorUserAccountId?: string;
  status?: DeploymentTaskStatus;
  queuedAt: string;
  startedAt?: string;
  completedAt?: string;
  processSnapshotId?: string;
  variableSnapshotId?: string;
  availability?: DeploymentCompatibilityAvailability;
  components: DeploymentTimelineComponent[];
  lastSuccessful: boolean;
  redeployAvailable: boolean;
}

export interface DeploymentTimelineComponent {
  key: string;
  name: string;
  type: ReleaseBundleComponentType;
  version: string;
}

export interface DeploymentTimelineCompareRef {
  taskId?: string;
  legacyDeploymentRevisionId?: string;
}

export interface DeploymentTimelineComparisonAvailability {
  process: boolean;
  steps: boolean;
  variables: boolean;
}

export interface DeploymentTimelineComparison {
  base: DeploymentTimelineItem;
  compare: DeploymentTimelineItem;
  process: DeploymentTimelineProcessChange;
  availability: DeploymentTimelineComparisonAvailability;
  components: DeploymentTimelineComponentChange[];
  steps: DeploymentTimelineStepChange[];
  variables: DeploymentTimelineVariableChange[];
}

export interface DeploymentTimelineProcessChange {
  baseProcessSnapshotId?: string;
  compareProcessSnapshotId?: string;
  baseRevisionNumber?: number;
  compareRevisionNumber?: number;
  baseCanonicalChecksum?: string;
  compareCanonicalChecksum?: string;
  changed: boolean;
}

export interface DeploymentTimelineComponentChange {
  key: string;
  name: string;
  kind: DeploymentTimelineChangeKind;
  baseVersion?: string;
  compareVersion?: string;
  baseType?: ReleaseBundleComponentType;
  compareType?: ReleaseBundleComponentType;
}

export interface DeploymentTimelineStepChange {
  stepKey: string;
  name: string;
  kind: DeploymentTimelineChangeKind;
  baseActionType?: string;
  compareActionType?: string;
  baseIncluded?: boolean;
  compareIncluded?: boolean;
}

export interface DeploymentTimelineVariableChange {
  key: string;
  kind: DeploymentTimelineChangeKind;
  baseStatus?: string;
  compareStatus?: string;
  baseSource?: string;
  compareSource?: string;
  baseRedacted: boolean;
  compareRedacted: boolean;
  baseValue?: unknown;
  compareValue?: unknown;
  baseReference?: string;
  compareReference?: string;
}

export interface DeploymentTimelineRedeploy {
  plan: DeploymentPlan;
  warning: string;
}

export interface DeploymentTask {
  id: string;
  createdAt?: string;
  updatedAt?: string;
  queuedAt: string;
  startedAt?: string;
  completedAt?: string;
  taskType?: string;
  deploymentPlanId?: string;
  deploymentPlanTargetId?: string;
  deploymentTargetId?: string;
  applicationId?: string;
  releaseBundleId?: string;
  channelId?: string;
  environmentId?: string;
  actorUserAccountId?: string;
  status: DeploymentTaskStatus;
  queueOrder?: number;
  stepRuns: DeploymentStepRun[];
}

export interface DeploymentStepRun {
  id: string;
  createdAt?: string;
  updatedAt?: string;
  startedAt?: string;
  completedAt?: string;
  taskId: string;
  deploymentPlanId?: string;
  deploymentPlanStepId?: string;
  stepKey: string;
  name: string;
  actionType: string;
  status: DeploymentStepRunStatus;
  sortOrder: number;
  skippedReason?: string;
}

export interface DeploymentTaskTimeline {
  organizationId: string;
  taskId: string;
  events: DeploymentStepRunEvent[];
}

export interface DeploymentStepRunEvent {
  id: string;
  createdAt: string;
  occurredAt: string;
  organizationId?: string;
  taskId: string;
  stepRunId: string;
  taskLeaseId?: string;
  agentId?: string;
  sequence: number;
  type: DeploymentStepRunEventType;
  message?: string;
  progressPercent?: number;
  details?: Record<string, unknown>;
  redacted: boolean;
  logs: DeploymentStepRunLog[];
  outputs: DeploymentStepRunOutput[];
}

export interface DeploymentStepRunLog {
  id?: string;
  createdAt?: string;
  occurredAt?: string;
  eventId?: string;
  stream: 'stdout' | 'stderr' | 'system';
  severity: 'debug' | 'info' | 'warn' | 'error';
  body: string;
  redacted: boolean;
}

export interface DeploymentStepRunOutput {
  id?: string;
  name: string;
  value?: unknown;
  sensitive: boolean;
  redacted: boolean;
}

File: api/deployment_timeline.go
package api

import (
	"encoding/json"
	"time"

	"github.com/distr-sh/distr/internal/types"
	"github.com/google/uuid"
)

type DeploymentTimelineQueryRequest struct {
	ApplicationID          *uuid.UUID `query:"applicationId"`
	ReleaseBundleID        *uuid.UUID `query:"releaseBundleId"`
	EnvironmentID          *uuid.UUID `query:"environmentId"`
	DeploymentTargetID     *uuid.UUID `query:"deploymentTargetId"`
	CustomerOrganizationID *uuid.UUID `query:"customerOrganizationId"`
	Limit                  int        `query:"limit"`
	IncludeNonTerminal     bool       `query:"includeNonTerminal"`
}

type DeploymentTimelineCompareQueryRequest struct {
	BaseTaskID                        *uuid.UUID `query:"baseTaskId"`
	BaseLegacyDeploymentRevisionID    *uuid.UUID `query:"baseLegacyDeploymentRevisionId"`
	CompareTaskID                     *uuid.UUID `query:"compareTaskId"`
	CompareLegacyDeploymentRevisionID *uuid.UUID `query:"compareLegacyDeploymentRevisionId"`
}

type DeploymentTimeline struct {
	Items []DeploymentTimelineItem `json:"items"`
}

type DeploymentTimelineItem struct {
	Source                     types.DeploymentTimelineItemSource        `json:"source"`
	TaskID                     uuid.UUID                                 `json:"taskId"`
	LegacyDeploymentID         uuid.UUID                                 `json:"legacyDeploymentId,omitempty"`
	LegacyDeploymentRevisionID uuid.UUID                                 `json:"legacyDeploymentRevisionId,omitempty"`
	SyntheticReleaseID         uuid.UUID                                 `json:"syntheticReleaseId,omitempty"`
	DeploymentPlanID           uuid.UUID                                 `json:"deploymentPlanId"`
	DeploymentPlanTargetID     uuid.UUID                                 `json:"deploymentPlanTargetId"`
	DeploymentTargetID         uuid.UUID                                 `json:"deploymentTargetId"`
	ApplicationID              uuid.UUID                                 `json:"applicationId"`
	ApplicationName            string                                    `json:"applicationName"`
	ReleaseBundleID            uuid.UUID                                 `json:"releaseBundleId"`
	ReleaseNumber              string                                    `json:"releaseNumber"`
	ChannelID                  uuid.UUID                                 `json:"channelId"`
	ChannelName                string                                    `json:"channelName"`
	EnvironmentID              uuid.UUID                                 `json:"environmentId"`
	EnvironmentName            string                                    `json:"environmentName"`
	CustomerOrganizationID     *uuid.UUID                                `json:"customerOrganizationId,omitempty"`
	DeploymentTargetName       string                                    `json:"deploymentTargetName"`
	ActorUserAccountID         *uuid.UUID                                `json:"actorUserAccountId,omitempty"`
	Status                     types.TaskStatus                          `json:"status"`
	QueuedAt                   time.Time                                 `json:"queuedAt"`
	StartedAt                  *time.Time                                `json:"startedAt,omitempty"`
	CompletedAt                *time.Time                                `json:"completedAt,omitempty"`
	ProcessSnapshotID          *uuid.UUID                                `json:"processSnapshotId,omitempty"`
	VariableSnapshotID         *uuid.UUID                                `json:"variableSnapshotId,omitempty"`
	Availability               types.DeploymentCompatibilityAvailability `json:"availability"`
	Components                 []DeploymentTimelineComponent             `json:"components"`
	LastSuccessful             bool                                      `json:"lastSuccessful"`
	RedeployAvailable          bool                                      `json:"redeployAvailable"`
}

type DeploymentTimelineComponent struct {
	Key     string                           `json:"key"`
	Name    string                           `json:"name"`
	Type    types.ReleaseBundleComponentType `json:"type"`
	Version string                           `json:"version"`
}

type DeploymentTimelineComparison struct {
	Base       DeploymentTimelineItem              `json:"base"`
	Compare    DeploymentTimelineItem              `json:"compare"`
	Process    DeploymentTimelineProcessChange     `json:"process"`
	Components []DeploymentTimelineComponentChange `json:"components"`
	Steps      []DeploymentTimelineStepChange      `json:"steps"`
	Variables  []DeploymentTimelineVariableChange  `json:"variables"`
}

type DeploymentTimelineProcessChange struct {
	BaseProcessSnapshotID    *uuid.UUID `json:"baseProcessSnapshotId,omitempty"`
	CompareProcessSnapshotID *uuid.UUID `json:"compareProcessSnapshotId,omitempty"`
	BaseRevisionNumber       int        `json:"baseRevisionNumber,omitempty"`
	CompareRevisionNumber    int        `json:"compareRevisionNumber,omitempty"`
	BaseCanonicalChecksum    string     `json:"baseCanonicalChecksum,omitempty"`
	CompareCanonicalChecksum string     `json:"compareCanonicalChecksum,omitempty"`
	Changed                  bool       `json:"changed"`
}

type DeploymentTimelineComponentChange struct {
	Key            string                             `json:"key"`
	Name           string                             `json:"name"`
	Kind           types.DeploymentTimelineChangeKind `json:"kind"`
	BaseVersion    string                             `json:"baseVersion,omitempty"`
	CompareVersion string                             `json:"compareVersion,omitempty"`
	BaseType       types.ReleaseBundleComponentType   `json:"baseType,omitempty"`
	CompareType    types.ReleaseBundleComponentType   `json:"compareType,omitempty"`
}

type DeploymentTimelineStepChange struct {
	StepKey           string                             `json:"stepKey"`
	Name              string                             `json:"name"`
	Kind              types.DeploymentTimelineChangeKind `json:"kind"`
	BaseActionType    string                             `json:"baseActionType,omitempty"`
	CompareActionType string                             `json:"compareActionType,omitempty"`
	BaseIncluded      *bool                              `json:"baseIncluded,omitempty"`
	CompareIncluded   *bool                              `json:"compareIncluded,omitempty"`
}

type DeploymentTimelineVariableChange struct {
	Key              string                             `json:"key"`
	Kind             types.DeploymentTimelineChangeKind `json:"kind"`
	BaseStatus       types.VariableResolutionStatus     `json:"baseStatus,omitempty"`
	CompareStatus    types.VariableResolutionStatus     `json:"compareStatus,omitempty"`
	BaseSource       types.VariableResolutionSource     `json:"baseSource,omitempty"`
	CompareSource    types.VariableResolutionSource     `json:"compareSource,omitempty"`
	BaseRedacted     bool                               `json:"baseRedacted"`
	CompareRedacted  bool                               `json:"compareRedacted"`
	BaseValue        json.RawMessage                    `json:"baseValue,omitempty"`
	CompareValue     json.RawMessage                    `json:"compareValue,omitempty"`
	BaseReference    string                             `json:"baseReference,omitempty"`
	CompareReference string                             `json:"compareReference,omitempty"`
}

type DeploymentTimelineRedeploy struct {
	Plan    DeploymentPlan `json:"plan"`
	Warning string         `json:"warning"`
}
